jupyter_ydoc/ynotebook.py

```diff
@@ -245,26 +249,79 @@ def set(self, value: Dict) -> None:
                 "id": str(uuid4()),
             }
         ]
+        old_ycells_by_id = {ycell["id"]: ycell for ycell in self._ycells}
 
         with self._ydoc.transaction():
-            # clear document
-            self._ymeta.clear()
-            self._ycells.clear()
+            try:
+                new_cell_list: List[tuple[Map, dict]] = []
+                retained_cells = set()
+
+                # Determine cells to be retained
+                for new_cell in new_cells:
+                    cell_id = new_cell.get("id")
+                    if cell_id and (old_ycell := old_ycells_by_id.get(cell_id)):
+                        old_cell = self._cell_to_py(old_ycell)
+                        if old_cell == new_cell:
+                            new_cell_list.append((old_ycell, old_cell))
+                            retained_cells.add(cell_id)
+                            continue
+                    # New or changed cell
+                    new_cell_list.append((self.create_ycell(new_cell), new_cell))
+
+                # First delete all non-retained cells
+                if not retained_cells:
+                    # fast path if no cells were retained
+                    self._ycells.clear()
+                else:
+                    index = 0
+                    for old_ycell in list(self._ycells):
+                        if old_ycell["id"] not in retained_cells:
+                            self._ycells.pop(index)
+                        else:
+                            index += 1
+
+                # Now add new cells
+                index = 0
+                for new_ycell, new_cell in new_cell_list:
+                    if len(self._ycells) > index:
+                        # we need to compare against a python cell to avoid
+                        # an extra transaction on new cells which are not yet
+                        # integrated into the ydoc document.
```

Comment:
You could use `new_ycell.prelim.get("id")` and avoid storing the python cells, or only store python cells and create ycells when needed?